GREL Zip function

sunilnatraj opened the pull request:
Zip function addition to GREL

Fixes #5440 

Changes proposed in this pull request:
- Zip function takes in a 2 or more Array parameters and generates output which is an Array of arrays 
- The length of the output will be that of the shortest input parameter
- Supports text, number, date
- Usage - zip(array 1, array 2, ...)
- Examples
1. zip( [1, 2], ["A", "B"] ) produces [ [ 1, "A" ], [ 2, "B" ] ]
2. zip( [1, 2], ["A", "B"] , [ now(), now() ] ) produces  [ 1, "A", "2024-08-01T09:24:02.063008Z" ], [ 2, "B", "2024-08-01T09:24:02.063017Z" ] ]

sunilnatraj commented:
@wetneb @ostephens @jquartel Please review the PR for zip function

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Override
+    public String getDescription() {
+        return FunctionDescription.arr_join();
```
This looks like it's going to pull in the wrong description

sunilnatraj replied:
Updated description

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Override
+    public String getParams() {
+        return "array a, string sep";
```
This parameter description appear left over from the Join function and needs to be updated.

sunilnatraj replied:
Updated param description

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Override
+    public String getReturns() {
+        return "string";
```
This return also needs to be updated to match the new behavior.

sunilnatraj replied:
Updated return description

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,153 @@
+        return "string";
+    }
+
+    private static ArrayList collectElements(Object a, Object b) {
+        ArrayList returnList = new ArrayList<>();
```
It's good practice to use interfaces rather than concrete classes where possible.

```suggestion
    private static List collectElements(Object a, Object b) {
        List returnList = new ArrayList<>();
```

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,153 @@
+            }
+            return output;
+        }
+        return new EvalError(EvalErrorMessage.expects_one_array(ControlFunctionRegistry.getFunctionName(this)));
```
This error message doesn't look correct since the function was spec'd to accept multiple arrays.

sunilnatraj replied:
Updated error message and handled all scenarios

tfmorris commented on main/tests/server/src/com/google/refine/expr/functions/arrays/ZipTests.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Test
+    public void zipArray() throws ParsingException {
+        String[] test = { "[ [1,2,3], ['A','B','C'] ].zip()", "[[1, A], [2, B], [3, C]]" };
```
The tests need to be updated to the following form which matches the original specification
```suggestion
        String[] test = { "zip([1,2,3],['A','B','C'])", "[[1, A], [2, B], [3, C]]" };
```

sunilnatraj replied:
Updated all test cases

tfmorris reviewed (changes requested):
Thanks for the contribution! I didn't realize it until I was well into the review, so some of the earlier comments may not make sense, but the implementation doesn't match the specification.

What you wrote:

> Zip function takes in a 2 or more Array parameters and generates output which is an Array of arrays

is correct, but then the usage example later shows just a single array parameter:

> `[a, b, c, d]`.zip()

when I would expect `zip(a, b, c, d)` (or `a.zip(b, c, d)` in the postfix notation version).

sunilnatraj commented:
@tfmorris Thank you for reviewing the code and feedback. I have addressed all the points. 

1. zip( [1, 2], ["A", "B"] ) produces [ [ 1, "A" ], [ 2, "B" ] ]
2. zip( [1, 2], ["A", "B"] , [ now(), now() ] ) produces [ 1, "A", "2024-08-01T09:24:02.063008Z" ], [ 2, "B", "2024-08-01T09:24:02.063017Z" ] ]

ostephens commented:
I don't know the answer to this, but just asking the question... is it correct that this should be implemented as a function, or should it be a control? Most code we have that iterate over an array are done as controls, although I've never really fully understood the reasons. I asked this question a long time ago on the (then) forum: https://groups.google.com/g/openrefine-dev/c/QvMSiHu-8u0/m/HcZt7PioAQAJ

@tfmorris can you comment?

tfmorris commented:
@sunilnatraj Thanks for the update. I'll have another look.

@ostephens David's comment on the Control interface says

> * Interface of GREL controls such as if, forEach, forNonBlank, with. A control can decide which part of the code to
> * execute and can affect the environment bindings. Functions, on the other hand, can't do either.

which I think summarizes the main difference. For a function, all the arguments get evaluated and if they generate errors, those will propagate, preventing the execution of the function (which wouldn't work for controls like `ifError`).

Since this is a straightforward transformation without any conditional execution or evaluation, I think implementing it as a Function instead of a Control is the right thing.

tfmorris commented on main/tests/server/src/com/google/refine/expr/functions/arrays/ZipTests.java:
```diff
@@ -0,0 +1,48 @@
+    @Test
+    public void zipArray() throws ParsingException {
+        String[] test = { "zip([1,2,3], ['A','B','C'])", "[[1, A], [2, B], [3, C]]" };
+        parseEval(bindings, test);
```
`parseEval` uses `toString` internally, so I don't really trust it for testing this kind of operation. I think it would be better to have your own help which knows how to compare two arrays of arrays to make sure that you're getting the results that you expect.

sunilnatraj replied:
@tfmorris I have fixed the incorrect type conversion and validated it via the application. 

I will check on updating the unit test to do array comparison.

sunilnatraj replied:
@tfmorris Added test cases with support to compare the output array against expected result.

tfmorris reviewed (changes requested):
I'll do some more testing and review, but one thing I noticed with my interactive testing is that `zip([1,2,3],[4,5,6])` results in 
    `[ [ 1, "4" ], [ 2, "5" ], [ 3, "6" ] ]` 
instead of the correct 
    `[ [ 1, 4 ], [ 2, 5 ], [ 3, 6 ] ]`

This is being masked in the tests because of the operation of the `parseEval` helper.

I think you need a helper which knows how to compare the actual values, not their string equivalents.

ostephens commented:
Looks like this is getting close to being merged.  Note that the official documentation will need updating with an explanation of this function - if this PR gets merged please create an Issue in https://github.com/OpenRefine/openrefine.org to add the documentation

tfmorris reviewed (changes requested):
This is greatly improved and produces the expected results. Thanks for all your work.

The main thing left to improve is to try and simplify the code as much as possible. I left some specific suggestions for various pieces, but I think the main body of the function needs to be restructured. 

I would suggest breaking the processing into two phases:
1. Validate each argument and collect some type of Stream/iterator/spliterator for it (I used spliterators for my experiments)
2. Generate output tuples (lists) from the next elements of each input until one of the inputs is empty

I think this will allow simple direct construction of the output without touching the inputs multiple times and will be easier to understand.

sunilnatraj commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,159 @@
+
+            for (int i = 0; i < args.length; i++) {
+                Object v = args[i];
+                if (v == null) {
+                    return new EvalError(
+                            EvalErrorMessage.expects_at_least_two_or_more_array_args(ControlFunctionRegistry.getFunctionName(this)));
+                }
+
+                if (!(v.getClass().isArray() || v instanceof List<?> || v instanceof ArrayNode)) {
+                    return new EvalError(
+                            EvalErrorMessage.expects_at_least_two_or_more_array_args(ControlFunctionRegistry.getFunctionName(this)));
+                }
```
merged the 2 checks

sunilnatraj commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,159 @@
+                }
+
+                if (v.getClass().isArray()) {
+                    ziparg2.addAll(Arrays.asList((Object[]) v));
+                } else if (v instanceof ArrayNode) {
+                    ArrayNode a = (ArrayNode) v;
+                    if (a.isArray()) {
+                        for (JsonNode objNode : a) {
+                            ziparg2.add(objNode);
+                        }
+                    }
+                } else {
+                    for (Object z : ExpressionUtils.toObjectList(v)) {
+                        if (z != null) {
+                            ziparg2.add(z);
+                        } else {
+                            ziparg2.add(z);
+                        }
+                    }
+                }
+                if (isFirst) {
+                    for (Object x : ziparg2) {
+                        ziparg1 = new ArrayList() {
+
+                            {
+                                add(x);
+                            }
+                        };
+                        output.add(ziparg1);
+                    }
+                } else {
+                    output = (List<List>) Streams.zip(output.stream(),
+                            ziparg2.stream(),
+                            (a, b) -> collectElements(a, b))
+                            .collect(Collectors.toList());
+                }
+                isFirst = false;
+                ziparg2.clear();
+            }
+            return output;
```
@tfmorris I have refactored the flow  - process the input arguments and then generate output
I did not replace the zip function from google streams as internally it is also using streams to merge the 2 inputs.  Please review the updated flow and if you believe that the use to google streams zip needs to be avoided.

sunilnatraj commented:
@tfmorris Thank you the review and feedback. 

I have refactored the flow - process the input arguments and then generate output
I did not replace the zip function from google streams as internally it is also using streams to merge the 2 inputs. Please review the updated flow and if you believe that the use to google streams zip needs to be avoided.

sunilnatraj commented:
@tfmorris Hi Can you review the PR when you get a chance

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,134 @@
+                input.add(ziparg);
+            }
+
+            output = (List<List>) Streams.zip(input.get(0).stream(),
+                    input.get(1).stream(),
+                    (a, b) -> collectElements(a, b))
+                    .collect(Collectors.toList());
+
+            if (input.size() > 2) {
+                for (int i = 2; i < input.size(); i++) {
+                    output = (List<List>) Streams.zip(output.stream(),
+                            input.get(i).stream(),
+                            (a, b) -> collectElements(a, b))
+                            .collect(Collectors.toList());
+                }
+            }
+            return output;
```
Then you can change the order of iteration so that you first iterate by element and then by input argument similar to the following:

```suggestion
            List<List> result = new ArrayList<>();
            boolean done = false;
            while (!done) {
                List element = new ArrayList<>();
                
                // Construct an output element by getting 
                // input element from each iterator
                // if any iterator is empty, we are done

                if (!done) {
                    result.add(element);
                }
            }
           return result;
```

Does that make sense? Let me know if you have any questions.

sunilnatraj replied:
@tfmorris Thanks for the feedback. Can you also confirm if the output of the zip is not limited to the shortest input. 
Is the below acceptable output of zip function?

-  zip([1,2], ['A','B','C'], ['X','Y','Z']) -> [[[1, A, X], [2, B, Y], [null, C, Z]]]

tfmorris replied:
Is there a reason that you'd prefer your alternative? If there's a user needs based reason, we should discuss it on the original issue where folks will see the discussion, but if it's just an implementation based preference, I think it would be better to make the implementation conform to the specification as currently given, i.e. the output for your example should be

`zip([1,2], ['A','B','C'], ['X','Y','Z']) -> [[[1, A, X], [2, B, Y]]]`

sunilnatraj replied:
@tfmorris The query came up as part of implementation. No change required in the zip working definition

tfmorris commented on main/resources/com/google/refine/grel/FunctionDescription.properties:
```diff
@@ -6,6 +6,7 @@ arr_join=Joins the items in the array with sep, and returns it all as a string.
 arr_reverse=Reverses array a.
 arr_sort=Sorts the array in ascending order. Sorting is case-sensitive, uppercase first and lowercase second.
 arr_uniques=Returns the array with duplicates removed. Case-sensitive.
+arr_zip=Combine two or more arrays into a single array of arrays, where the elements at the same index from the input arrays are paired together to form a sub-array in the output array.
```
```suggestion
arr_zip=Combines two or more arrays into a single array of arrays, where the elements at the same index from the input arrays are grouped together to form a sub-array in the output array.
```

tfmorris reviewed (changes requested):
Thanks. That's definitely improved, but I think we can simplify it even more to get rid of the N>2 vs N=2 distinction so that all inputs are processed uniformly. See my suggestions inline and let me know if you have any questions.

Thanks for the sticking with this and apologies for the delay in reviewing your updates.

sunilnatraj commented:
@tfmorris Thanks for your inputs, I have removed the Streams zip dependency and updated the working as per your feedback. Please review the PR.

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,115 @@
+            }
+
+            List<List> output = new ArrayList<>();
+            List<Object> currentElements = new ArrayList<>(minSize);
+
+            for (int i = 0; i < minSize; i++) {
+                currentElements.clear();
```
```suggestion
            boolean done = false;
            while (!done)  {
                List<Object> currentElements = new ArrayList<>();
```

I think it's simpler to just create your output list inside the loop. You are going to use it every time except the last iteration, so there's really no appreciable extra object creation overhead.

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,115 @@
+                }
+            }
+
+            int minSize = Integer.MAX_VALUE;
+            for (Spliterator<?> iterator : iterators) {
+                minSize = Math.min(minSize, (int) iterator.estimateSize());
+            }
+
```
```suggestion
```

The fact that the method is called `estimateSize()` is a red flag that it probably shouldn't be relied upon for correct operation of the algorithm.

I would suggest iterating until one of the iterators is empty.

sunilnatraj replied:
Removed the size estimation logic

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,115 @@
+                } else if (v.getClass().isArray()) {
+                    iterators.add(Arrays.stream((Object[]) v).spliterator());
+                } else if (v instanceof ArrayNode) {
+                    ArrayNode a = (ArrayNode) v;
+                    List<Object> ziparg = new ArrayList<>();
+                    if (a.isArray()) {
+                        for (JsonNode objNode : a) {
+                            ziparg.add(objNode);
+                        }
+                    }
+                    iterators.add(ziparg.stream().spliterator());
```
```suggestion
                    iterators.add(((ArrayNode) v).spliterator());
```

This can be simplified by using the `spliterator` method of ArrayNode.

sunilnatraj replied:
I added the loop as the iterator fails when processing Array node.

tfmorris replied:
It doesn't fail for me using your tests. Do you have a separate test which demonstrates the failure? Could you add it to the test suite please so that someone doesn't try to optimize this in the future. Thanks!

tfmorris commented on main/src/com/google/refine/expr/functions/arrays/Zip.java:
```diff
@@ -0,0 +1,115 @@
+            for (int i = 0; i < minSize; i++) {
+                currentElements.clear();
+                for (Spliterator<?> iterator : iterators) {
+                    Object[] elementHolder = new Object[1];
+                    if (iterator.tryAdvance(e -> elementHolder[0] = e)) {
+                        currentElements.add(elementHolder[0]);
+                    }
+                }
+                output.add(new ArrayList<>(currentElements));
```
```suggestion
                    if (!iterator.tryAdvance(e -> currentElements.add(e)) {
                        done = true;
                    }
                }
                if (!done) {
                    output.add(currentElements);
                }
```

You can add the element to your output list directly inside the tryAdvance method and by inverting the if test, you can set a `done` flag when one of your iterators is empty and it's time to leave.

tfmorris reviewed (changes requested):
This is looking really good! I think it's almost ready to merge. Hopefully all my individual comments/suggestions make sense when all stitched together, but basically:

- don't depend on Spliterator.estimateSize. I don't think it's guaranteed to give you a true size (thus the "estimate" in the name).
- a few simplifications to small sections of the code.

I think it'll be ready to merge after this round. Thanks for all your work!

sunilnatraj commented:
@tfmorris Thank you for your inputs; I have further simplified the logic. I hope it is now good to be committed

tfmorris commented on main/tests/server/src/com/google/refine/expr/functions/arrays/ZipTests.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Test
+    public void testZipforParams() {
```
```suggestion
    public void testZipForParams() {
```

tfmorris commented on main/tests/server/src/com/google/refine/expr/functions/arrays/ZipTests.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Test
+    public void testZipwithStrArrays() {
```
```suggestion
    public void testZipWithStrArrays() {
```

tfmorris commented on main/tests/server/src/com/google/refine/expr/functions/arrays/ZipTests.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Test
+    public void testZipwithIntArrays() {
```
```suggestion
    public void testZipWithIntArrays() {
```

tfmorris commented on main/tests/server/src/com/google/refine/expr/functions/arrays/ZipTests.java:
```diff
@@ -0,0 +1,109 @@
+        List arg1 = List.of(1, 2, 3);
+        List arg2 = List.of(7.89, 8.90, 9.01);
+
+        List<List> expected = List.of(
+                List.of(1, 7.89),
+                List.of(2, 8.90),
+                List.of(3, 9.01));
+
```
```suggestion
```

This is unused

tfmorris reviewed (commented):
Looks great! It's ready for merge except for the question I have below on the rejected suggestion for the `ArrayNode.spliterator`. I can't get it to fail for me, so a test would be good to make sure someone doesn't shoot themselves in the foot in the future.

Thanks for sticking with it and sorry the iterations have taken so long with me being away.

sunilnatraj commented:
@tfmorris The unit test that was failing, it was case for testing JSON. I updated the code and re-ran all the tests and no failure noted. Thank you for all your guidance on this.


`// Tests for JSON array
        String[] test8 = {
                "zip([\"Bob has a cat & dog\", \"Cat's name is Dotty\", \"Dog's name is Dots\"], [11, 22, 33, 44, 55], '[\"Doe\", \"Foe\", 456, 789]'.parseJson() )",
                "[[Bob has a cat & dog, 11, \"Doe\"], [Cat's name is Dotty, 22, \"Foe\"], [Dog's name is Dots, 33, 456]]" };
        parseEval(bindings, test8);`

tfmorris reviewed (approved):
Looks great. Thank you! And congratulations on another PR approved.

sunilnatraj commented:
Thanks @tfmorris
